### DISCUSSION TITLE
Asynchronous serialization

### DISCUSSION THREAD

**Discussion by sogartar** [nlohmann/json#2799] [Q&A]
I would like to serialize/deserialize asynchronously to/from stream of bytes. This would avoid blocking reads/writes. Looking at the documentation it seems that the API is geared towards a synchronous approach. What would be a proper way to do it asynchronously?

What I could come up with is to have a stream that hides the async details with coroutines. Then `stream >> j;` will not block the thread and also avoid expensive context switches. This seems like a lot of work. Is there a better way?

**Comment by nlohmann** [comment]
Do you mean something like a StAX parser (https://de.wikipedia.org/wiki/Streaming_API_for_XML)?

**Reply by sogartar** [reply]
I am not familiar with StAX, but from the description it seems that it is doing and asynchronous event driven parsing.

My desire is to use [Boost ASIO](https://www.boost.org/doc/libs/1_76_0/doc/html/boost_asio.html)'s async TCP [streams](https://www.boost.org/doc/libs/1_76_0/doc/html/boost_asio/overview/core/streams.html).
There instead of blocking the reading thread to complete a reading/writing operation it utilizes an IO Executor that uses 
OS's facilities to wait on multiple sockets for available data/space in the kernel buffers. Then you can use less threads to service many sockets and avoid doing thread context switches.
Basically, it is a one giant loop that does
1. Wait for possible IO on all sockets.
2. Read/Write data from/to sockets that are ready for IO.

**Reply by nlohmann** [reply]
This library does not support such an approach.

**Reply by sogartar** [reply]
Is there some stateful Parser in the API, that can parse from a buffer? The use case is to read some data from the stream into a buffer then feed the buffer to the parser and repeat until the json is parsed?

**Reply by nlohmann** [reply]
This may work using streams, but I have not check this yet.

**Reply by maddanio** [reply]
I have been mulling it. the ability to do streaming parsing may be the thing that tips me towards using boost json instead of nlohmann. but I have been wondering, with c++20 coroutines it would actually be relatively easy to do: basically one would have to turn parse (or parse_sax_internal which is the actual parser) into a co-routine when supplied with a co-routine getting string chunks. i.e. the ia.get_character() would be a co-routine. then to supply the non-coroutine version one would make an adapter that never yields and unpack the result. looks to me like it would not be much of a descign problem, just lots of mechanical steps of adding co_await, co_return etc

**Reply by maddanio** [reply]
I think that is the only sensible way short of rewriting the parser. because basically the task of turning the current parser into a streaming parser would mean to completely "invert" it from pull-based to push-based. and I think using co-routines would be much better, because structurally the parser would remain the same

**Reply by maddanio** [reply]
this was surprisingly easy. I managed to turn the lexer into a coroutine here:
https://github.com/maddanio/json/tree/feature/coroutines
it has only the lexer test enabled and it passes. its currently a bit hacked, you need to update submodules to get cppcoro, build it, and hack the built library path into jsons cmakelist. also the coroutine type is hard-coded to cpproro::task right now, but that should not be a problem. next I want to get the parser test(s) also working.

**Reply by maddanio** [reply]
the biggest challenge is to support non-coroutine compilers. I suppose this will mean using a lot of macro magic to only optionally insert co_await/co_return keywords.

**Reply by maddanio** [reply]
I think this is actually pretty cool, beacause when also using the sax parser directly it means one never has to store the entire thing at all, just parse a stream as it comes and emit events.

**Reply by nlohmann** [reply]
Awesome! I will check it out during the weekend!

**Reply by maddanio** [reply]
got the parser as well (minus the binary stuff which I ifdefed out). might also some time plug in the maintained fork of cppcoro which supports cmake to make it easier.
right now doesnt do anything interesting, but one could for example use asios coroutine support:
https://www.boost.org/doc/libs/1_78_0/doc/html/boost_asio/overview/core/cpp20_coroutines.html

**Reply by maddanio** [reply]
the intersting functions are suffixed _coro, the other ones just sync wait on the coro versions

**Reply by nlohmann** [reply]
Nice! The changes seem to be really small. Did you make benchmarks already showing the upsides of going asynchronously? Furthermore, any plans to make this independent of a third-party library?

**Reply by maddanio** [reply]
I don't expect any performance upsides actually. It just enables exploiting underlying asynchronicity from networking for example. I would already be happy if all tests pass and performance is unchanged :).
what is required from the third party lib is a basic awaitable type and a sync_wait function. Maybe if licenses allow we could take just those parts from cppcoro for example. and then we would need to enable backward compatibility with pre-20 compilers and maybe improve the interface a bit.

**Reply by maddanio** [reply]
ok, I finally cleaned it up a bit, tests pass, but already the regular tests take way longer, so it seems the compiler fails to inline completely through the co-routines. so I would say it is very likely not the right way to go about this. so to the OP: if you want to feed your parser with chunks as they come (what is what you want in an async context) you probably wanna head over to boost::json, which in some ways seems like a more modern version of this (and probably heavily inspired by this).
